### ISSUE TITLE
How will we save users' grammar and options?

### ISSUE THREAD

**Issue by BramVanroy** [BramVanroy/earley-interface#5]
The easiest, local approach would be to use [localStorage](https://developer.mozilla.org/nl/docs/Web/API/Window/localStorage) on the front-end. However, this won't be ideal from a usability perspective: people often switch devices and always need access to their grammars. I think we'll most definitely need to move to the back-end for this one.

I'm not familiar with the options here, nor what's the most efficient considering storing page and load times. Maybe we could do create a button on the page that "saves" the current form values (i.e. the whole grammar and all the options) into a database, which is then linked to a new, randomly generated url? Similar to how [JSFiddle](http://jsfiddle.net/) does it when you click "Save".

To save space, we might also automatically delete entries after a defined period, e.g. three months.

**Comment by garfix** [comment]
The JSFiddle option is very nice. It is easy to make and we can store the user's config either in a database or as JSON in separate files. The latter is probably best since no database is needed, and the structure does not lend itself for database storage very well.

**Comment by BramVanroy** [comment]
@garfix would you automatically delete these files after x time, or not implement that (yet)? Also, the JSON option seems best indeed!

**Comment by garfix** [comment]
Yes, I was thinking about deleting them a year after they were last used.

**Comment by BramVanroy** [comment]
@garfix How would a generated JSON file then be linked to a **unique(!)** url? In the front-end you can find the hash in the url, and then load a JSON file with the same string as the hash. (http://.../XYZ -> XYZ.json) Maybe there's a better way in the back-end?

**Comment by garfix** [comment]
This is my idea. When the user first presses 'save':

1. the php app creates a random string. i.e. 'cv244cbbb242'
2. the grammar and settings are stored as JSON in a file with the random string in the name (cv244cbbb242.json)
3. the app performs an HTTP redirect to the Earley page with the random string in the url (just like Fiddle does) (http://domain/earley-interface?conf=cv244cbbb242)

When the Earley page is loaded with the conf id in the url (http://domain/earley-interface?conf=cv244cbbb242):

1. the php app locates the JSON file with the id in the name (cv244cbbb242.json)
2. the contents of the file is used to populate the forms

and

The action of the 'save' button now also contains the random string. When the user presses 'save', the php app replaces the existing JSON file with the new JSON structure reflecting the grammar and settings.

**Comment by BramVanroy** [comment]
Love the idea! Instead of creating a *new* random string each time, maybe we could do things more similar to JSFiddle's approach:

On first save, create a new folder on the server (random) e.g. cv244cbbb242, and create a file in that folder named cv244cbbb242-**1**.json. Then, each time the save button is clicked, add a new file to that folder, incrementing the trailing number. The URL would then look like this: http://domain/earley-interface?conf=cv244cbbb242/2 for the *second version*. This way, a user can always go back to a previous version they have build (and saved) before.

**Comment by garfix** [comment]
Aha, I did not know that. Yes, that's even better!

Is this ok for url?
?conf=abcdef  (for the first version)
?conf=abcdef&v=2 (for the second version)
etc

We could also do /abcdef/2 as url but this would require some .htaccess work, and we don't know if the environment where this application will finally run supports it.

**Comment by BramVanroy** [comment]
@garfix I think it's safe to assume that .htaccess is available. If I'm right, the script shouldn't be anything else than:

 ````
RewriteEngine On
RewriteBase /
RewriteRule (\w+)[/](\d+)$ /?conf=$1&v=$2 [L,NC]
```

However, I'm not sure that if we use this approach (which I'd like, but isn't *needed*) you can easily grab the information you need to load the correct configuration/json. Maybe [this](http://stackoverflow.com/a/5493174/1150683) answer is useful for you/to give you ideas? Because as you know I'm not a hero in this stuff.

**Comment by BramVanroy** [comment]
The URL rewrite is fixed since https://github.com/BramVanroy/earley-interface/commit/a787322847f7eb836c1c5a821db1c4d61b034561
